### PR TITLE
[optim] rprop: handle complex params as independent real params

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref #65711

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -908,6 +907,13 @@ def test_rprop(self):
                     lr=2e-4, maximize=maximize),
                 constructor_accepts_maximize=True
             )
+            self._test_complex_2d(optimizer)
+            self._test_complex_optimizer(
+                lambda param: optimizer([param], lr=0.001)
+            )
+            self._test_complex_optimizer(
+                lambda param: optimizer([param], lr=0.001, maximize=True)
```
Comment (kshitij12345) [inline test/test_optim.py:915]: Add test with `maximize` argument for previously supported tensor.

File: test/test_optim.py [hunk 2]
```diff
@@ -311,10 +311,9 @@ def _test_complex_optimizer(self, optimizer_constructor):
         complex_opt = optimizer_constructor(complex_param)
         real_opt = optimizer_constructor(real_param)
 
-        for i in range(3):
+        for _ in range(3):
             complex_param.grad = torch.randn_like(complex_param)
-            real_param.grad = torch.view_as_real(complex_param.grad)
-
+            real_param.grad = torch.view_as_real(complex_param.grad.clone())
```
Comment (kshitij12345) [inline test/test_optim.py:316]: grad gets mutated internally by the optimizer.

Comment (albanD) [reply test/test_optim.py:316]: What? Which optimizer does that??

File: torch/optim/rprop.py [hunk 3]
```diff
@@ -110,7 +110,12 @@ def step(self, closure=None):
                 if len(state) == 0:
                     state['step'] = 0
                     state['prev'] = torch.zeros_like(p, memory_format=torch.preserve_format)
-                    state['step_size'] = grad.new().resize_as_(grad).fill_(group['lr'])
+                    if p.dtype.is_complex:
+                        # Complex Number should be as if they are two independent real numbers.
+                        # Hence the step_size shouldn't be zero for imaginary part.
+                        state['step_size'] = grad.new().resize_as_(grad).fill_(complex(group['lr'], group['lr']))
```
Comment (albanD) [inline torch/optim/rprop.py:116]: nit: why not keep the step_size real and just make it twice as big by using `.resize_as_(torch.view_as_real(grad))...`

**Review by albanD — COMMENTED** [review]
Just one question about testing. Sounds ok otherwise.

File: test/test_optim.py (hunk 2 above)
Comment (kshitij12345) [reply test/test_optim.py:316]: rprop multi tensor path:
https://github.com/pytorch/pytorch/blob/eff28d61c9c961e6c2724b78f57441ee2e3e40cb/torch/optim/rprop.py#L225-L226

File: torch/optim/rprop.py (hunk 3 above)
Comment (kshitij12345) [reply torch/optim/rprop.py:116]: Down the line there are inplace operations with step_size
https://github.com/pytorch/pytorch/blob/eff28d61c9c961e6c2724b78f57441ee2e3e40cb/torch/optim/rprop.py#L197-L198

Here sign is complex tensor viewed as real and `step_size` could turn out to be different for real and imaginary part after a few steps.

File: test/test_optim.py (hunk 2 above)
Comment (kshitij12345) [reply test/test_optim.py:316]: Not intended probably 🤔

Comment (albanD) [reply test/test_optim.py:316]: Ho wow I missed that :o
Yes this should be out of place!
Can you fix it here please?

File: torch/optim/rprop.py (hunk 3 above)
Comment (albanD) [reply torch/optim/rprop.py:116]: Sure, let's ignore that.

**Review by albanD — APPROVED** [review]
Thanks for the fix!

**Pull request closed without merging**